### PR TITLE
add ec2 instance support to aws v2

### PR DESCRIPTION
**Pull request by jjlimepoint**
This builds upon the aws v2 work in here to allow for actually building ec2 instances.  also includes small chef13 fixes, since that's what i test on...

Fixes creation of:
ec2 instance
ebs volume
vpc
subnet
security groups
elb

### CODE REVIEW

File: lib/chef/provisioning/aws_driver/driver.rb [hunk 1]
```diff
@@ -768,7 +778,7 @@ def cloudsearch(api_version="20130101")
     end
 
     def ec2
-      @ec2 ||= ::Aws::EC2.new(config: aws_config)
+      @ec2 ||= ::Aws::EC2::Client.new(config: aws_config)
```
Comment (harikesh-kolekar) [inline lib/chef/provisioning/aws_driver/driver.rb]: @jjlimepoint , I got an error: `ArgumentError: invalid configuration option :config` while trying to access `config: aws_config`. Can you please try with the line below:
```
@ec2 ||= ::Aws::EC2::Client.new(aws_config)
```

**Comment by jjlimepoint** [comment]
Oh yeah.... don't know how I missed that one - oddly, that code path is never hit for me now (I made the change when it was being hit), and actually anything which attempts to use it won't actually work I suspect.

**Comment by jjlimepoint** [comment]
Updated this to correct the checking of subnets - for some reason this started being necessary, though it was working before - weird.

**Comment by NimishaS** [comment]
@jjlimepoint , Thanks for contribution. We'll test this and merge.

**Comment by NimishaS** [comment]
@jjlimepoint , can you please mention the resources that you are fixing in the description of the PR?

**Comment by NAshwini** [comment]
Hi @jjlimepoint , thanks for your contribution.

While testing the code I stumbled upon some issues:
1) In file `lib/chef/provider/aws_vpc.rb`:
```
   a)(byebug) retry_with_backoff(::Aws::EC2::Errors::InvalidVpcID::NotFound)do end
*** NameError Exception: uninitialized constant Aws::EC2::Errors::InvalidVpcID::NotFound
```
```
   b) (byebug)  vpc.tags['Name'] = new_resource.name
*** TypeError Exception: no implicit conversion of String into Integer
```
2) Faced same issues in file `lib/chef/provider/aws_internet_gateway.rb` also.

3) `aws_dhcp_option` resource is not working as only half of the changes are made. Current changes are only in `lib/chef/resource/aws_dhcp_options.rb`, however changes are required in `lib/chef/provider/dhcp_option.rb` too.

@jjlimepoint, `aws_dhcp_option` , `aws_vpc` , `aws_internet_gateway` resources are already implemented in [harry/use_AWS_SDK_v2](https://github.com/chef/chef-provisioning-aws/pull/530) branch. Can you please rebase your branch with that?
We are working on other resources as well. We'll be able to merge this PR if the updated resources are functional end to end.

**Comment by jjlimepoint** [comment]
rebased and update - the following now pass usaing v2 api - tested with docs/examples/aws_sg.rb and docs/examples/aws_ebs_volume :

ec2 instance
ebs volume
vpc
subnet
security groups
elb

**Comment by NAshwini** [comment]
Hi @jjlimepoint,

While running network_interface recipe I am facing issue for creation of network_interface.
error are as follows:
```
(byebug) eni = new_resource.driver.ec2_resource.create_network_interface(options)
[2017-06-15T18:30:50+05:30] INFO: [Aws::EC2::Client 200 2.700884 0 retries] describe_subnets(subnet_ids:["subnet-2f05544a"])


*** ArgumentError Exception: parameter validator found 3 errors:
  - missing required parameter params[:subnet_id]
  - unexpected value at params[:subnet]
  - unexpected value at params[:security_groups]

nil
```
```
(byebug) retry_with_backoff(::Aws::EC2::Errors::InvalidNetworkInterfaceID::NotFound)do end
*** NameError Exception: uninitialized constant Aws::EC2::Errors::InvalidNetworkInterfaceID::NotFound

nil
(byebug)
```

-We wil fix it now.

**Pull request merged**